text/0000-socket-timeouts.md

```diff
@@ -0,0 +1,142 @@
+indefinitely (i.e., until the connection is closed or data is
+transferred). But there are many contexts in which timing out a
+blocking call is important.
+
+The [goal of the current IO system][io-reform] is to gradually expose
+cross-platform, blocking APIs for IO, especially APIs that directly
+correspond to the underlying system APIs. Sockets are widely available
+with nearly identical system APIs across the platforms Rust targets,
+and this includes support for timeouts via [sockopts][sockopt].
+
+So timeouts are well-motivated and well-suited to `std::net`.
+
+# Detailed design
+
+The proposal is to *directly expose* the timeout functionality
+provided by [`setsockopt`][sockopt], in much the same way we currently
+expose functionality like `set_nodelay`:
+
+```rust
+impl TcpStream {
+    pub fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn read_timeout(&self) -> Option<Duration>;
+
+    pub fn set_write_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn write_timeout(&self) -> Option<Duration>;
+}
+
+impl UdpSocket {
+    pub fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn read_timeout(&self) -> Option<Duration>;
+
+    pub fn set_write_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn write_timeout(&self) -> Option<Duration>;
+}
+```
+
+The setter methods take an amount of time in the form of a `Duration`,
+which is [undergoing stabilization][duration-reform]. They are
+implemented via straightforward calls to `setsockopt`. The `Option` is
+used to signify no timeout (for both setting and
+getting). Consequently, `Some(Duration::new(0, 0))` is a possible
+argument; the setter methods will return an IO error of kind
+`InvalidInput` in this case. (See Alternatives for other approaches.)
+
+The corresponding socket options are `SO_RCVTIMEO` and `SO_SNDTIMEO`.
+
+# Drawbacks
+
+One potential downside to this design is that the timeouts are set
+through direct mutation of the socket state, which can lead to
+composition problems. For example, a socket could be passed to another
+function which needs to use it with a timeout, but setting the timeout
+clobbers any previous values. This lack of composability leads to
+defensive programming in the form of "callee save" resets of timeouts,
+for example. An alternative design is given below.
+
+The advantage of binding the mutating APIs directly is that we keep a
+close correspondence between the `std::net` types and their underlying
+system types, and a close correspondence between Rust APIs and system
+APIs. It's not clear that this kind of composability is important
+enough in practice to justify a departure from the traditional API.
+
+# Alternatives
+
+## Taking `Duration` directly
+
+Using an `Option<Duration>` introduces a certain amount of complexity
+-- it raises the issue of `Some(Duration::new(0, 0))`, and it's
```

Comment:
`Some(Duration::new(0, 0))` can be mostly back-compatibly be mapped to mark the socket non-blocking at a later time, though.